### DISCUSSION TITLE
Module visibility criteria

### DISCUSSION THREAD

**Discussion by matchai** [starship/starship#52] [General]
> Pulling in a conversation we briefly had on Discord which was started by @youssefhabri

As we've been reimplementing modules from spaceship and spacefish, we've been wondering whether the criteria for a module to appear were still appropriate.

The particular situation @youssefhabri encountered was that they were in their home directory with `.hyper.js` in it. As a result, the Node.js module was shown. Does that make sense for us?

I always found it a little strange that for the Node.js module, the version would be visible if we had a `*.js`, `package.json`, _or_  a `node_modules` folder. Would we _want_ to show the node section if there is a folder with just `node_modules` but no `package.json`? Does that make sense to do?

Any thoughts on how we can better identify user intent from context? 🤔

**Comment by Snuggle** [comment]
I personally feel like we should never show anything just because of a file-extension existing in $PWD sounds a little silly. The rest, I'm not sure.

**Comment by jletey** [comment]
@matchai I feel in this case we could just have *.js && (node_modules || package.json) or even just *.js && package.json

**Comment by matchai** [comment]
> I feel in this case we could just have *.js && (node_modules || package.json) or even just *.js && package.json

Unfortunately, that would make it so projects, where source files are in a `src/` directory, would not show the module. 🤔

**Comment by Multimo** [comment]
checking the folder for a file extension is a bug that i introduced with my refactor. #64 fixes the issue.

An interesting question would be should we fallback, in the case of scanning the current path, to looking into the parent directories or associate the type of the project (Rust vs Python vs JS) with the github repo?

What do you guys think?

**Comment by matchai** [comment]
> An interesting question would be should we fallback, in the case of scanning the current path, to looking into the parent directories or associate the type of the project (Rust vs Python vs JS) with the github repo?

Ooh! I like the sound of that.
If a user goes into a new folder within a Node project, the Node version is still relevant. 😄 

We could eventually add some smart caching around that to reduce the amount of I/O needed when you're in the same project. For example, as long as the git workdir hasn't changed, and the `package.json` last modified date hasn't changed, we can safely assume that the package version is the same as it was in the last prompt. No need to reread `package.json`.

**Comment by pseudo-su** [comment]
I think it'd make perfect sense to only show it if the **current directory** or **any parent directories** contained a `package.json`.

I can only speak for myself but I personally can't think of any situation where I'd want the starfish node module to appear where the above wasn't the case.

*Technically* just because a folder has a `.js` file in it doesn't mean that it uses nodejs. it *could* be:
* Code only intended to be run in a browser without any nodejs tooling
* A config file that happens to be `.js` (eg .hyper.js)
* An alternative JS/TS runtime like [deno](https://github.com/denoland/deno)

**Comment by pseudo-su** [comment]
This is probably a different discussion but a feature I think would be cool to put in the starship's node module would be to read the `package.json` [engines config](https://docs.npmjs.com/files/package.json#engines) and display some kind of warning/color change if the current `node` or `npm` version doesn't satisfy the requirements specified by the project

**Comment by matchai** [comment]
> I think it'd make perfect sense to only show it if the current directory or any parent directories contained a package.json.

I completely agree. I would probably limit it to the current directory or any parent directory until the git root of the project (if there is one).

In that same vein, I think we should rethink quite a few of our modules associated with programming languages in the same way. What files actually represent that a certain runtime will be used?

> I think would be cool to put in the starship's node module would be to read the package.json engines config and display some kind of warning/color change if the current node or npm version doesn't satisfy the requirements specified by the project.

Yes!!! 👍 That's a great idea.

**Comment by g2p** [comment]
Personally I think files starting with a dot should be ignored when using extension-based filename detection.
(this doesn't apply to well-known files starting with a dot — like .travis.yml for example — but I don't think any are used at the moment)

**Comment by youssefhabri** [comment]
@g2p maybe the hidden files could be configurable/opt-in per-module?

If a module needs it, it can enable it, if not, by default, don't check to hidden files.

**Comment by g2p** [comment]
@youssefhabri
That can be added if and when a need for globbing hidden files comes up.
I think globbing has a lot of potential for slowdowns and non-user-obvious consequences (as in #280), which is why, if it is used, it should be limited to things that ls will show.
